### PR TITLE
Implement Color refinement

### PR DESCRIPTION
**Pull request by pdelvo**

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+                    return false;
+                }
+                
+                // check surjectivity, i.e. are the colors in the set {1, ..., maximumColor} used?
+                if(alpha.getColorClasses().size() != alpha.getNumberColors()) {
+                    return false;
+                }
+                
+                for(int color = 1; color <= alpha.getNumberColors(); color++) {
+                    if(!alpha.getColors().values().contains(color)) {
```
Comment (pdelvo) [inline jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java]: This algorithm is currently O(n^3)

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+     * @param alpha the surjective l-coloring to be checked
+     * @param graph the graph that is colored by alpha
+     * @return whether alpha is a valid surjective l-coloring for the given graph
+     */
+    private boolean isAlphaConsistent(Coloring<V> alpha, Graph<V, E> graph) {
+        // check if every vertex is assigned a color
+        if(alpha.getColors().size() != graph.vertexSet().size()) {
+            for(V v : alpha.getColors().keySet()) {
+                // TODO: What does this do? Maybe check surjectivity of alpha?
+                if(alpha.getColors().get(v) <= alpha.getNumberColors() && alpha.getColors().get(v) >= 1) {
```
Comment (pdelvo) [inline jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java]: `alpha.getColors().get(v)` can be stored in a variable

**Comment by Watercrystal** [comment]
Update: I added some tests for CR and they fail; we are going to have to hunt some bugs I guess.

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java (hunk 2 above)
Comment (christophgruene) [reply jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java]: #wearedumb... lol. I fixed that

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java (hunk 1 above)
Comment (christophgruene) [reply jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java]: #wearedumb... lol. I fixed that

**Comment by christophgruene** [comment]
I close this PR as there are unecessary commits that I cannot delete.

**Pull request closed without merging**